build.rs

```diff
@@ -0,0 +1,165 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate regex;
+
+use regex::Regex;
+use std::env;
+use std::fs::read_dir;
+use std::fs::File;
+use std::io::{Read, Write};
+use std::process::Command;
+
+fn main() {
+    // This build script creates files in the `src` directory. Since that is
+    // outside Cargo's OUT_DIR it will cause an error when this crate is used
+    // as a dependency. Therefore, the user must opt-in to regenerating the
+    // Rust files.
+    if env::var_os("CARGO_FEATURE_REGENERATE").is_none() {
+        println!("cargo:rerun-if-changed=build.rs");
+        return;
+    }
+
+    let file_names: Vec<_> = read_dir("proto")
+        .expect("Couldn't read proto directory")
+        .map(|e| {
+            format!(
+                "proto/{}",
+                e.expect("Couldn't list file").file_name().to_string_lossy()
+            )
+        })
+        .collect();
+    let file_names: Vec<_> = file_names.iter().map(|s| &**s).collect();
+
+    for f in &file_names {
+        println!("cargo:rerun-if-changed={}", f);
+    }
+
+    match BufferLib::from_env_vars() {
+        BufferLib::Prost => {
+            unimplemented!("Prost support is not yet implemented");
+        },
+
+        BufferLib::Protobuf => {
+            check_protoc_version();
+
+            generate_protobuf_files(file_names);
+            let mod_names: Vec<_> = read_dir("src/rsprotobuf")
+                .expect("Couldn't read src directory")
+                .filter_map(|e| {
+                    let file_name = e.expect("Couldn't list file").file_name();
+                    file_name
+                        .to_string_lossy()
+                        .split(".rs")
+                        .next()
+                        .map(|n| n.to_owned())
+                })
+                .collect();
+            replace_read_unknown_fields(&mod_names);
+            generate_protobuf_rs(&mod_names);
+        },
+    }
+}
+
+#[derive(Eq, PartialEq)]
+enum BufferLib {
+    Prost,
+    Protobuf,
+}
+
+impl BufferLib {
+    fn from_env_vars() -> BufferLib {
+        match (
+            env::var_os("CARGO_FEATURE_LIB_PROST"),
+            env::var_os("CARGO_FEATURE_LIB_RUST_PROTOBUF"),
+        ) {
+            (Some(_), Some(_)) | (None, None) => {
+                panic!("You must use exactly one of `lib-rust-protobuf` and `prost-buf` features")
+            }
+            (Some(_), _) => BufferLib::Prost,
+            (_, Some(_)) => BufferLib::Protobuf,
+        }
+    }
+}
+
+fn check_protoc_version() {
```

Comment:
It's kind of unsatisfying to need to add a bash script for this purpose. This wholesale breaks our compatibility with systems without Bash like Windows. Do you think we can use `Command::new` and get the version from the `build.rs`?

Reply:
It's just some regex matching, maybe completely rewrite with Rust.

Reply:
Please do. I prefer it if this was part of the PR.

Reply:
Give me some time to practice rust regex.